tests/main/interfaces-audio-playback-record/task.yaml

```diff
@@ -0,0 +1,104 @@
+
+# Only classic Ubuntu has the pulseaudio mediation patches
+systems: [ ubuntu-1*-*64 ]
+
+environment:
+    PLAY_FILE: "/snap/test-snapd-audio-record/current/usr/share/sounds/alsa/Noise.wav"
+    # today, only 19.04 and 19.10 have a mediating pulseaudio
+    EXFAIL: "ubuntu-1[468]"
+
+prepare: |
+    #shellcheck source=tests/lib/pkgdb.sh
+    . "$TESTSLIB"/pkgdb.sh
+    snap install --edge test-snapd-audio-record
+
+    apt-get update
+    apt-get install -y pulseaudio pulseaudio-utils
+
+    echo "Create XDG_RUNTIME_DIR=/run/user/12345"
+    # shellcheck disable=SC2174
+    mkdir -m 700 -p /run/user/12345 || true
+    chown test:test /run/user/12345
+
+    # ensure we have a clean pulse directory
+    test -d /home/test/.config && mv /home/test/.config /home/test/.config.spread
+    mkdir -m 700 /home/test/.config
+    mkdir -m 700 /home/test/.config/pulse
+    chown test:test /home/test/.config /home/test/.config/pulse
+
+restore: |
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --kill" test || true
+    snap remove test-snapd-audio-record
+    apt-get autoremove --purge -y pulseaudio pulseaudio-utils
+    rm -rf /run/user/12345 /home/test/.config/pulse
+    if [ -d /home/test/.config.spread ]; then
+        rm -rf /home/test/.config
+        mv /home/test/.config.spread /home/test/.config
+    fi
+
+execute: |
+    # we need -p so that XDG_RUNTIME_DIR is honored, but that preserves HOME,
+    # so specify it too since pulseaudio fails to start otherwise
+    echo "Start pulseaudio"
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "pulseaudio --exit-idle-time=30" test &
+
+    echo "Then wait for the socket to show up"
+    retry-tool -n 10 test -S /run/user/12345/pulse/native
+
+    echo "Check pulseaudio"
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 retry-tool -n 10 su -p -c "pulseaudio --check" test
+
+    echo "The unconfined user can play audio"
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "/usr/bin/paplay $PLAY_FILE" test
+
+    echo "The unconfined user can record audio"
+    HOME=/home/test XDG_RUNTIME_DIR=/run/user/12345 su -p -c "/snap/test-snapd-audio-record/current/bin/parec-simple" test
+
+    echo "The audio-playback interface is connected by default"
+    snap interfaces -i audio-playback | MATCH ":audio-playback .*test-snapd-audio-record"
+
+    echo "The audio-record interface is disconnected by default"
+    snap interfaces -i audio-record | MATCH "\- +test-snapd-audio-record:audio-record"
+
+    echo "When the audio-playback plug is connected"
+    snap connect test-snapd-audio-record:audio-playback
+
+    echo "Then the snap can play audio"
+    su -l -c "test-snapd-audio-record.play $PLAY_FILE" test
+
+    echo "When the audio-record plug is connected"
+    snap connect test-snapd-audio-record:audio-record
+
+    echo "Then the snap can record audio"
+    su -l -c "test-snapd-audio-record.recsimple" test
```

Comment:
should we add a test before audio-record is connected that shows there are apparmor denials when the plug is not connected and then check that the app doesn't have any apparmor denials when the plug is connected?

Reply:
> should we add a test before audio-record is connected that shows there are apparmor denials when the plug is not connected and then check that the app doesn't have any apparmor denials when the plug is connected?

This is not something we do in other interface tests, largely because kernel rate limiting is problematic. We've historically only checked the access and if the test (positive or negative) fails, we see in the logs the audit messages.